### PR TITLE
db: Switch to statement rather than row trigger

### PR DESCRIPTION
**Pull request by ryanslade**
The trig_soft_delete_orphan_repo_by_external_service_repo was previously
triggered per row when an external_service_repos row was deleted.

We have switched this to run after the entire statement affecting
external_service_repos has run. Running per row led to a quadratic number
of statments being run.

Onc significant side effect of this change is that after any row is
deleted in external_service_repos ANY orphaned repos will be soft
deleted rather than repos only affected by the deleted row. The reason
for this is that after statement triggers don't have access to what
changed in Postgre 9.6 which we need to support. From 10 we could use
transition tables.

Given that whenver we add a repo we should ALWAYS also add an entry in
the external_service_repos table this should not pose an issue.

Performance:

In testing locally, deleting a an external service with a large number
of repos dropped from around 20s to 8s. When the number of repos owned
by the external service was doubled, the runtime also doubled showing
that we are now O(n).

Closes: https://github.com/sourcegraph/sourcegraph/issues/16069

### CODE REVIEW

File: migrations/frontend/1528395755_alter_orphaned_repo_trigger.up.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    -- performs a clean up to soft-delete orphan repositories.
+    UPDATE
+        repo
+    SET
+        name = soft_deleted_repository_name(name),
+        deleted_at = transaction_timestamp()
+    WHERE
+      deleted_at IS NULL
+      AND id NOT IN (
+        SELECT DISTINCT(repo_id) FROM external_service_repos
```
Comment (tsenart) [inline migrations/frontend/1528395755_alter_orphaned_repo_trigger.up.sql:20]: #### Query plan for this query

```
Update on repo  (cost=0.43..250136501159.34 rows=1933371 width=534)
  ->  Seq Scan on repo  (cost=0.43..250136501159.34 rows=1933371 width=534)
        Filter: ((deleted_at IS NULL) AND (NOT (SubPlan 1)))
        SubPlan 1
          ->  Materialize  (cost=0.43..124213.57 rows=2052123 width=4)
                ->  Unique  (cost=0.43..105935.95 rows=2052123 width=4)
                      ->  Index Only Scan using external_service_repos_repo_id_external_service_id_unique on external_service_repos  (cost=0.43..96158.01 rows=3911176 width=4)
```

#### Query plan for this alternative

```sql
UPDATE
    repo
SET
    name = soft_deleted_repository_name(name),
    deleted_at = transaction_timestamp()
WHERE
    deleted_at IS NULL AND
    NOT EXISTS (
        SELECT FROM external_service_repos WHERE repo_id = repo.id
    );
```

```
Update on repo  (cost=13.68..1787234.81 rows=1815167 width=540)
  ->  Merge Anti Join  (cost=13.68..1787234.81 rows=1815167 width=540)
        Merge Cond: (repo.id = external_service_repos.repo_id)
        ->  Index Scan using repo_pkey on repo  (cost=0.43..702676.08 rows=3866743 width=525)
              Filter: (deleted_at IS NULL)
        ->  Index Scan using external_service_repos_repo_id_external_service_id_unique on external_service_repos  (cost=0.43..573407.01 rows=3911176 width=10)
```

Can you check that it runs faster? Even if it doesn't in practice, the plan is better, so I think we should use it.

**Review by unknwon — APPROVED** [review]
Approve with @tsenart's comment.

**Review by asdine — APPROVED** [review]
👍🏼

File: migrations/frontend/1528395755_alter_orphaned_repo_trigger.up.sql (hunk 1 above)
Comment (ryanslade) [reply migrations/frontend/1528395755_alter_orphaned_repo_trigger.up.sql]: Nice, in my testing the new query is twice as fast ⚡

**Pull request merged**